aggregator/leader_flush_mgr.go

```diff
@@ -0,0 +1,341 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package aggregator
+
+import (
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	schema "github.com/m3db/m3aggregator/generated/proto/flush"
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3x/clock"
+	"github.com/m3db/m3x/instrument"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/sync"
+
+	"github.com/uber-go/tally"
+)
+
+type leaderFlushManagerMetrics struct {
+	queueSize         tally.Gauge
+	flushTimesPersist instrument.MethodMetrics
+}
+
+func newLeaderFlushManagerMetrics(scope tally.Scope) leaderFlushManagerMetrics {
+	return leaderFlushManagerMetrics{
+		queueSize:         scope.Gauge("queue-size"),
+		flushTimesPersist: instrument.NewMethodMetrics(scope, "flush-times-persist", 1.0),
+	}
+}
+
+type leaderFlushManager struct {
+	sync.RWMutex
+
+	nowFn                  clock.NowFn
+	checkEvery             time.Duration
+	jitterEnabled          bool
+	workers                xsync.WorkerPool
+	flushTimesKeyFmt       string
+	flushTimesStore        kv.Store
+	flushTimesPersistEvery time.Duration
+	logger                 xlog.Logger
+	scope                  tally.Scope
+
+	rand                *rand.Rand
+	flushTimes          flushMetadataHeap
+	flushTimesKey       string
+	lastPersistAt       time.Time
+	flushedSincePersist bool
+	flushTask           *leaderFlushTask
+	metrics             leaderFlushManagerMetrics
+}
+
+func newLeaderFlushManager(opts FlushManagerOptions) roleBasedFlushManager {
```

Comment:
How much do the different FlushManagers differ by? If it's not much, can you combine them into one file with some toggles for leader etc?